Feature/add helpdesk functionality

courtneymyers opened the pull request:
Implements the Helpdesk functionality.

@dpgalloway there are a few places you'll need to do the verification of the user's EPA WAA groups and update the code appropriately. I'll flag those individually in comments below.

courtneymyers commented on app/client/src/components/app.tsx:
```diff
@@ -21,13 +21,29 @@ import Loading from "components/loading";
 
+type HelpdeskAccess = "idle" | "pending" | "success" | "failure";
+
+export function useHelpdeskAccess() {
+  const [helpdeskAccess, setHelpdeskAccess] = useState<HelpdeskAccess>("idle");
+
+  useEffect(() => {
+    setHelpdeskAccess("pending");
+    fetchData(`${serverUrl}/api/helpdesk-access`)
+      .then((res) => setHelpdeskAccess("success"))
+      .catch((err) => setHelpdeskAccess("failure"));
+  }, []);
+
+  return helpdeskAccess;
+}
```
This is where the verification request/response occurs the client to the server. I setup this as a custom hook as it's called in 1) the `<Dashboard />` component to determine if the button/link to the `/dashboard` route should be displayed, and 2) on the `/dashboard` page itself (in case a user visits there directly). On the `/dashboard` route itself, nothing is shown to the user until the API call confirms the user has access. If it responds that they don't they're redirected back the `/` route.

@dpgalloway nothing for you to do here – just wanted to point out how I set it up.

courtneymyers commented on app/client/src/routes/helpdesk.tsx:
```diff
@@ -0,0 +1,258 @@
+      data: null,
+    });
+
+  if (
+    epaUserData.status !== "success" ||
+    helpdeskAccess === "idle" ||
+    helpdeskAccess === "pending"
+  ) {
+    return <Loading />;
+  }
+
+  if (helpdeskAccess === "failure") {
+    navigate("/", { replace: true });
+  }
```
Here's what prevents the user from accessing this page if they shouldn't have access.

@dpgalloway again, just an FYI.

courtneymyers commented on app/server/app/routes/api.js:
```diff
@@ -14,14 +19,15 @@ const router = express.Router();
-
 router.use(ensureAuthenticated);
 
+// --- verification used to check if user has access to the /helpdesk route
+router.get("/helpdesk-access", (req, res) => {
+  // TODO: return 200 if user is a member of the correct EPA WAA groups, else
+  // send an error status code that the client app's fetch will catch (401?)
+  res.sendStatus(200);
+});
+
```
@dpgalloway this is intentionally in the server's `/routes/api` file. It needs the same `ensureAuthenticate` check (not an elevated check just for helpdesk users – more on that in a bit), as it's called to determine if a regular authorized user is also a helpdesk user.

Inside the callback is where you'll want to verify the user is in the correct EPA WAA groups. If they are, you can return a simple 200 status; if they're not, you should return an error status code (probably 401, but the client app doesn't actually care, as long as it's an HTTP error code – it just needs to be caught in the client app fetch's `catch` block and that'll indicate the user doesn't have access).

courtneymyers commented on app/server/app/routes/api.js:
```diff
@@ -36,10 +42,13 @@ router.get("/epa-data", (req, res) => {
 router.get("/sam-data", (req, res) => {
   getSamData(req.user.mail)
     .then((samUserData) => {
+      // TODO: set `helpdeskUser` to true if they're a member of the correct EPA WAA groups
+      const helpdeskUser = false;
```
@dpgalloway here you'll want to do that same check to determine if the user is a help desk member. If they are, set the `helpdeskUser` variable to true, else set it false, as it's used in the if statement below to allow helpdesk users without any SAM.gov data to still access the app.

I'm realizing now I should probably limit a help desk user's access to create new forms if they don't have any SAM.gov data, but I can do that as a separate PR.

courtneymyers commented on app/server/app/routes/help.js:
```diff
@@ -0,0 +1,69 @@
+const router = express.Router();
+
+// TODO: update authentication check to restrict these API calls to users in the correct EPA WAA groups
+router.use(ensureAuthenticated);
```
@dpgalloway I'd imagine you'll want to create a similar `ensureAuthenticated` middleware, specifically for help desk users that also checks their EPA WAA groups. You can apply it to this entire file (as done here) as these routes are only called by help desk users.

dpgalloway commented:
@courtneymyers I believe my part is done to tie this all together. I added "ensureHelpdesk" middleware to check against all /help routes and the /api/helpdesk-access route. I added a "getComboKeys" function that uses the existing "getSamData", converts to an array of combo keys, and plugged that into the two spots you comments on in the API routes. Please double check when you can and merge if all looks good (or we can chat tomorrow AM if there's anything you want to discuss first).

courtneymyers commented:
@dpgalloway awesome, works great! The help desk restricted access works great based on the WAM groups you're checking from the JWT. The one question I have is on the naming of those groups. In your code you use an underscore (`csb_admin` and `csb_helpdesk`). In your user config file in the "saml-docker" repo you use hyphens instead of underscores (`csb-admin` and `csb-helpdesk`). It doesn't matter to me which is used, I just want to make sure we're matching what Tarsha named it – you may already be doing that here, and I can just update the groups in the "saml-docker" repo, but I just wanted to confirm which to use.

Also, thanks for slipping the `getComboKeys` in those other two places too – we're finally fetching forms only associated with the user! I saw one of the SAM.gov records associated with my email address has yours and Brad's email address associated with it too (I know that was the whole point of setting up the test data that way, so thank you). On my test project I added the hidden BAP combo keys field and added Brad as a user in my local copy of the config file in the "saml-docker" repo. I added two form submissions in my test project – one using a SAM.gov records that only has my email address associated with it, and another that has Brad's as well. Both showed on my dashboard as expected. I logged out from my email address, and logged in with Brad's and can confirm I see _only_ the submission that he should have access to. We'll do some more thorough testing of course, but I can tentatively confirm this works as expected!

Copying @coobr01 here.

courtneymyers commented:
@dpgalloway I'm going to merge this in. Let's just regroup on the EPA group names when you can – we can update them in either place as needed (either in this project, or the local SAML Docker project).
